File: perl/day17_part1_2021.pl

use strict;
use warnings;

# Read the target area from input.txt
open my $fh, '<', 'input.txt' or die "Cannot open input.txt: $!";
my $line = <$fh>;
close $fh;

# Parse the target area
my ($x_min, $x_max, $y_min, $y_max) = $line =~ /x=(\d+)\.\.(\d+), y=(-?\d+)\.\.(-?\d+)/;

my $max_height = 0;

# Iterate through possible initial velocities
for my $vx (0 .. $x_max) {
    for my $vy ($y_min .. abs($y_min)) {
        my ($x, $y) = (0, 0);
        my ($current_vx, $current_vy) = ($vx, $vy);
        my $current_max_height = 0;

        # Simulate the probe's trajectory
        while ($x <= $x_max && $y >= $y_min) {
            $x += $current_vx;
            $y += $current_vy;
            $current_max_height = $y if $y > $current_max_height;

            # Check if within target area
            if ($x >= $x_min && $x <= $x_max && $y >= $y_min && $y <= $y_max) {
                $max_height = $current_max_height if $current_max_height > $max_height;
                last;  # Found a valid trajectory, no need to continue
            }

            # Update velocities
            $current_vx-- if $current_vx > 0;
            $current_vy--;
        }
    }
}

print "The highest y position reached is: $max_height\n";

File: training_data/reflection-tuning/day17_part1_2021.py
import re

def solve(target):
    # Extract target area coordinates
    x_min, x_max, y_min, y_max = map(int, re.findall(r'-?\d+', target))
    
    # The highest initial y velocity that still hits the target
    # is the one that makes the probe reach the bottom of the target
    # in one step after returning to y=0
    max_y_velocity = abs(y_min) - 1
    
    # The highest point is reached when the initial upward velocity becomes zero
    max_height = (max_y_velocity * (max_y_velocity + 1)) // 2
    
    return max_height

# Read input from file
with open('input.txt', 'r') as file:
    target_area = file.read().strip()

# Solve and print the result
print(solve(target_area))
